From: Aurélien Gâteau <agateau@kde.org>
Date: Mon, 30 Apr 2012 10:21:15 +0000
Subject: Rework colors to avoid white-on-black information widgets and flat backgrounds
X-Git-Url: http://quickgit.kde.org/?p=kdelibs.git&amp;a=commitdiff&amp;h=678d7190dd27c0c0a7f7757cd16b1ec0febbdf96
---
Rework colors to avoid white-on-black information widgets and flat backgrounds
---


--- a/kdeui/widgets/kmessagewidget.cpp
+++ b/kdeui/widgets/kmessagewidget.cpp
@@ -218,55 +218,66 @@ KMessageWidget::MessageType KMessageWidg
     return d->messageType;
 }
 
+static void getColorsFromColorScheme(KColorScheme::BackgroundRole bgRole, QColor* bg, QColor* fg)
+{
+    KColorScheme scheme(QPalette::Active, KColorScheme::Window);
+    *bg = scheme.background(bgRole).color();
+    *fg = scheme.foreground().color();
+}
+
 void KMessageWidget::setMessageType(KMessageWidget::MessageType type)
 {
     d->messageType = type;
     KIcon icon;
-    KColorScheme::BackgroundRole bgRole;
-    KColorScheme::ForegroundRole fgRole;
-    KColorScheme::ColorSet colorSet = KColorScheme::Window;
+    QColor bg0, bg1, bg2, border, fg;
     switch (type) {
     case Positive:
         icon = KIcon("dialog-ok");
-        bgRole = KColorScheme::PositiveBackground;
-        fgRole = KColorScheme::PositiveText;
+        getColorsFromColorScheme(KColorScheme::PositiveBackground, &bg1, &fg);
         break;
     case Information:
         icon = KIcon("dialog-information");
-        bgRole = KColorScheme::NormalBackground;
-        fgRole = KColorScheme::NormalText;
-        colorSet = KColorScheme::Tooltip;
+        // There is no "information" background role in KColorScheme, use the
+        // colors of highlighted items instead
+        bg1 = palette().highlight().color();
+        fg = palette().highlightedText().color();
         break;
     case Warning:
         icon = KIcon("dialog-warning");
-        bgRole = KColorScheme::NeutralBackground;
-        fgRole = KColorScheme::NeutralText;
+        getColorsFromColorScheme(KColorScheme::NeutralBackground, &bg1, &fg);
         break;
     case Error:
         icon = KIcon("dialog-error");
-        bgRole = KColorScheme::NegativeBackground;
-        fgRole = KColorScheme::NegativeText;
+        getColorsFromColorScheme(KColorScheme::NegativeBackground, &bg1, &fg);
         break;
     }
-    const int size = KIconLoader::global()->currentSize(KIconLoader::MainToolbar);
-    d->iconLabel->setPixmap(icon.pixmap(size));
 
-    KColorScheme scheme(QPalette::Active, colorSet);
-    QBrush bg = scheme.background(bgRole);
-    QBrush border = scheme.foreground(fgRole);
-    QBrush fg = scheme.foreground();
+    // Colors
+    bg0 = bg1.lighter(110);
+    bg2 = bg1.darker(110);
+    border = KColorScheme::shade(bg1, KColorScheme::DarkShade);
+
     d->content->setStyleSheet(
         QString(".QFrame {"
-            "background-color: %1;"
+            "background-color: qlineargradient(x1: 0, y1: 0, x2: 0, y2: 1,"
+            "    stop: 0 %1,"
+            "    stop: 0.1 %2,"
+            "    stop: 1.0 %3);"
             "border-radius: 5px;"
-            "border: 1px solid %2;"
+            "border: 1px solid %4;"
             "}"
-            ".QLabel { color: %3; }"
+            ".QLabel { color: %5; }"
             )
-        .arg(bg.color().name())
-        .arg(border.color().name())
-        .arg(fg.color().name())
+        .arg(bg0.name())
+        .arg(bg1.name())
+        .arg(bg2.name())
+        .arg(border.name())
+        .arg(fg.name())
         );
+
+    // Icon
+    const int size = KIconLoader::global()->currentSize(KIconLoader::MainToolbar);
+    d->iconLabel->setPixmap(icon.pixmap(size));
 }
 
 QSize KMessageWidget::sizeHint() const

